subsys/net/lib/download_client/src/coap.c

```diff
@@ -45,11 +47,19 @@ int coap_block_update(struct download_client *client, struct coap_packet *pkt,
 			*blk_off);
 	}
 
+	memcpy(&received_block_ctx, &client->coap.block_ctx, sizeof(client->coap.block_ctx));
+
 	err = coap_update_from_block(pkt, &client->coap.block_ctx);
 	if (err) {
 		return err;
 	}
 
+	if (client->coap.block_ctx.current < received_block_ctx.current) {
```

Comment:
I suggest that we store the next expected block number instead (within `client->coap`) - that way we could detect duplicates, as well as out-of-order blocks. Please have a look at [lwm2m_engine](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_engine.c#L4257) where it's implemented that way.

Reply:
thanks!

Reply:
@rlubos , the next expected block is already calculated in coap_block_update (within coap_next_block). The goal is now to get the block from the current coap_packet. To get I need to implement a routine like this to subsys/net/lib/coap/coap.c within zephyr repo. Is this your intention? I´d go with that solution

`int coap_get_current_from_pkt(const struct coap_packet *cpkt)
{
	int block;
	size_t new_current;

	block = coap_get_option_int(cpkt, COAP_OPTION_BLOCK2);

	if (!is_request(cpkt)) {
		return -EINVAL;
	}

	new_current = GET_NUM(block) << (GET_BLOCK_SIZE(block) + 4);
	if (block == -ENOENT) {
		return 0;
	}

	if (GET_MORE(block)) {
		return -EINVAL;
	}

	if (GET_NUM(block) > 0 && GET_BLOCK_SIZE(block) != ctx->block_size) {
		return -EINVAL;
	}

	return new_current;
}`

Reply:
Yes, I was thinking of obtaining block number/data offset from the current coap packet, and compare it with the next expected one - only after that we update the block context.

I think we could start with a small helper function in the download_client iteself, if it's generic enough it could be put in the CoAP library upstream for reuse.

Reply:
done